### PR TITLE
feat: opportunitiesSlice opportunity-specific metadata

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

✅  ~~Blocked by https://github.com/shapeshift/lib/pull/1163~~ 

Adds a new `opportunitiesSlice.opportunitySpecific` union, discriminated on both `defiProvider` and `defiType` as T and U.

This will allow us to support savers cap, and at a higher-level, any arbitrary, chain-specific data we might need in the future.

See https://github.com/shapeshift/web/pull/3600 ~~NOTE: FOX LP will be broken with this refactor, given the fact that we're relying on `DefiProvider.FoxEthLP` at view-layer which was left over from previous implementations, but we don't need anymore now that we support `DefiType` as a LP/staking discriminator. Consider this branch WIP until I insert a branch under this one, removing `DefiProvider.FoxEthLP`.~~ 

TSC playgrounds:

- [Using an incompatible DefiProvider / DefiType combination](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBMAdgVwLZwCLAGYEsAqAnmMHAN4BQccAMrgI7K4AmuMhAChBADZwC8cAEQ8wQgDRU4ANQCGyHvEFCAbvMUSpAZRiyA1rkQBzAcIDOug8c3UAYrKipDJ5dgdPrk6vgh6kO-WdTIRhfJAB9C0DPCgBfCgpQSFgEFHQsPA4oCBUWYChyKQBJZh5SZRYymzgATWAHRGDCeqhEaq0AC1kSTtxsJWFO7uBe-urbCBAAURgOmg5ggFUAOSKZACZxyftHIOUhno6++B2PI2qAYQgzVGvgq5vr6oB5R7NcM2DX2-ezavwOtAAMZdQxaWQqfKfZT4AASzwAShdQY1wZCoH9JPEKMxgECeA5SOwSDJZDxkMBnthFohcBBEAAefDiOAAaQAfKZ8AgQDAkMxPshEHpEBAAO6IKQAfjZPL5iAFcD8hAg2Dg+Ck1Bl+AA2qyALqauAALjgQtxeEQwGYUlNVvRAG4KAB6ABUrrgAHFgPAyXxlZ9DGbafSABT4ACUcFkgpDjVdzpxeIJUCJxFIrOAhDMVJpdMZ+E5gm5oHliqFIvFjRlytV6pNcHt+SdbtdVA9F1Tsj5cFQ3TAQTrfqVWcDjSF+fDUbFbA6cDA2RIBWJpDU5OAZnb7bg4VFfNNAAMAD4jlVq-AHuCGXEgK+fd6oMA8QhwUKv4A8PgrsxAqC4MDwLM3b1h8cAWoYbAfi+shwGYyAAEavumcB1pm2a5nG4TtomuL4oSSEknm9IALL9s4TJFoU1B6lejRoTm1JxhRcAnrW576qacjrhh+ZMiyHJOrELbunAHC4ECegocggEdKQH7AKgSAwJ8YqkJWYpwNg0CwcgqbRgqvb6HJ6KELMQQfmYpAQAB+ZknACZJnhekrnARGIMR+RGMAzGCGJEkMm5pFgAOxgUSybHqpyABkUgcA4MC4GSAVxkFIVGBR7Its66pGY02DZOgyKyGCJBAn04kERuV6PmUimILoCX0iyZiyIpV64jBQ5gR8v64B43bWsG+bRswrCNYgdl4B+AoutlWkFDBGS4FkOR5FAzpLUQJJAhAqDwYY3b5i6IkAIKjZ8w6QGY7zwWUcBrhSnx1hOYaRtGF2NhufLMCh8EAFZ4vA5r5NGcAHiCxWIFopXlUCl7KvZiaOSmaYks8wXQDAE7sNDeKw3xrkstIlGBZ53lSCWvL8rG6mStQWr1qW1OnnW0hGgzixyszFaihK7P0zKnNMwqnwRdIOr4Ia9PSwLVEy-L1DWckWO0jjMN4ECnES-qOqLPqToKzL8SG9QprA5a1r86bn3ovzZsKjghiW9LdrALbmUUC56PK9jhC42VGsebozDdrIDIU1zIuYI7K25LiUBeK5keKpt6YUJR3uY77-v4xTieLInlDSzqS2x2tAB0ALAiiaJQhxcvF6nJDlwEVhGPXRcKy16JmFoyDBc+RSIPiyC4idQK-hSzC2IlMCmhYf7GAbXcQlCpEgH3A+EDP3bzzAi9GMv0tCUaJ-UPEHuJOAmOVXAmewL7QeyCHujh94yefKX2RxyDghf6t8d84f2jngLa5QQEEDTpRTu3RCB7wPsvGMlkYAlHgc4ZeC4AH5FNPgZeMAVA8DQUvKQK5TSLGXs6bKEwCjzBQhjB+qtcAbhZGZQMnxZikBam1GMr5ZJwBOtdH0JQpCULgNQ2ClhBz0JVmwJhgZy6kDFLJRoKpkASKMnAVRNCFhIJ9Cwo4bDeGkAEcgkoKE1QcNoDowR8B4LACCFEPwNpqCiIAOpHBBL2XARgOjwG0k4Hx8A9xwDsUY2CrVSA8MsSYoRzAWTAHLkYER2UAC0cAZyzCseotuLJVHZMiY0WhoQ-B0zgKItJGS5wlDKLkiAajHGRM0UwngzAUl2IcEEYpSBmoQDCcDKAz5nAxJQT9V0HwPTRJscIlxqT0mzjgA8H4cAtAYFZLU+plgiR8P6YM4w71kHdVTECGAz4el9IdgMwgQypmjPGWE4Z0zgzx12UYB5zAiGH2SeqPhbyUIFDeewgxdCfaqxfKmBcG4lKbhmWI7SRSwiIDMPo7sAByC6jRdHwDAMVKAXzxFKyzowqqMECRwGqcAZFMA0VhMxWEglDC2AvjFFAbo0Knn5BeQC00Op9ZfIBKQHZVy9mtTqfVcx7LLnXNMT9AADPpH6fZCChIAIzzhBqq+FJSWRBngjGAVqspCCqlT6BEh1rimi7Mwekz44A6gXs4Zq+9nD6hYnAS11qXx2qdcYXlML+VuuAGKBwP1MWAtAvSmR7AmnTSRQRcSfoXzKudOsP5cAADM+yfSfD7CAcufLngYGeKadG41JraV1RJINUBFQ7UfIdfaPBGUsngtJK88AzCAgUCGngQbswhNIBCyy9UpCpircwAFUpuW+p5AOVMzBJ0hO4GUWQiBl4TUUgu+1xDqA93zJu71nzqC6CMGYfdB8eWCTgFFO+0js7q3Ek-F+YdmSuUvjtRFgEvrFtvaCx9ocGT-x-lAcuizrgsibgk1uzhoFSFgaaIQW7ziJ0xag8wB7qiYKA6aQDFdQNIpIQQ+DiHqikIgWAlukjjCJ1EeI2hEbsZyP0aBUCliuGRMBcYm5eLtKOKkSC2RVVcAKPScozRdT8lid0lkzFTHDGTOleKyxtDaV2IcZs5xZTsruPEnOAJvjU16aCRAWxWzOERPevcm58TElfIqfM2hvGqOSYkyurJXTSnlLmZk8l6yXPRpaW0lonSEVnMsUa4wvyxlmAmT8rjMK7OZLw8s1ZvmGnnOeUKkwtLWBHJOYQUL2yLmcpufZO58nYmGqK5lt5RH0PUeyv6352kAW8PDT+xlAbB1Qu49o18CLY1ASpeizNWKcU9eBYSgTH1SU+d4aijjI26Xtajcy1llWMvGpGWYKd9Xvn6o28K24Qo-FqnC68krcqV0KtkEq0gqqlxwA1Qsdz2rGi6sskNGA62OXVZsaaxq23bUIbq2hg+QgXU8PfRYXb-rR3BpG2Gp6y2XxXJjSw4g8bPyJuTamjNobDK5vzYW79pa+DzRCbIStwbPi1uxQlBtTaQmtrYLBTtLTow9pu58UJXXh3UDh9Wrltr9SJySLgOdpo3A8EsonddwBatg8Tru+kCvnAkdkCe7lQhweJ3o6CnOGtTSd2oN3KEm8nyECHiPMeE8oBTx3nPUHavE4m9Xhide5vnwO9VzEc+WInQUCh7wBJPAIBGFDHyCwEYgA), with `chainSpecific` params being caught as incompatible
- [Using a compatible combination](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBMAdgVwLZwCLAGYEsAqAnmMHAN4BQccAMrgI7K4AmuMhAChBADZwC8cAEQ8wQgDRU4ANQCGyHvEFCAbvMUSpAZRiyA1rkQBzAcIDOug8c3UAYrKipDJ5dgdPrk6vgh6kO-WdTIRhfJAB9C0DPCgBfCgpQSFgEFHQsPA4oCBUWYChyKQBJZh5SZRYymzgATWAHRGDCeqhEaq0AC1kSTtxsJWFO7uBe-urbCBAAURgOmg5ggFUAOSKZACZxyftHIOUhno6++B2PI2qAYQgzVGvgq5vr6oB5R7NcM2DX2-ezavwOtAAMZdQxaWQqfKfZT4AASzwAShdQY1wZCoH9JPEKMxgECeA5SOwSDJZDxkMBnthFohcBBEAAefDiOAAaQAfKZ8AgQDAkMxPshEHpEBAAO6IKQAfjZPL5iAFcD8hAg2Dg+Ck1Bl+AA2qyALqauAALjgQtxeEQwGYUlNVvRAG4KAB6ABUrrgAHFgPAyXxlZ9DGbafSABT4ACUcFkgpDjVdzpxeIJUCJxFIrOAhDMVJpdMZ+E5gm5oHliqFIvFjRlytV6pNcHt+SdbtdVA9F1Tsj5cFQ3TAQTrfqVWcDjSF+fDUbFbA6cDA2RIBWJpDU5OAZnb7bg4VFfNNAAMAD4jlVq-AHuCGXEgK+fd6oMA8QhwUKv4A8PgrsxAqC4MDwLM3b1h8cAWoYbAfi+shwGYyAAEavumcB1pm2a5nG4TtomuL4oSSEknm9IALL9s4TJFoU1B6lejRoTm1JxhRcAnrW576qacjrhh+ZMiyHJOrELbunAHC4ECegocggEdKQH7AKgSAwJ8YqkJWYpwNg0CwcgqbRgqvb6HJ6KELMQQfmYpAQAB+ZknACZJnhekrnARGIMR+RGMAzGCGJEkMm5pFgAOxgUSybHqpyABkUgcA4MC4GSAVxkFIVGBR7Its66pGY02DZOgyKyGCJBAn04kERuV6PmUimILoCX0iyZiyIpV64jBQ5gR8v64B43bWsG+bRswrCNYgdl4B+AoutlWkFDBGS4FkOR5FAzpLUQJJAhAqDwYY3b5i6IkAIKjZ8w6QGY7zwWUcBrhSnx1hOYaRtGF2NhufLMCh8EAFZ4vA5r5NGcAHiCxWIFopXlUCl7KvZiaOSmaYks8wXQDAE7sNDeKw3xrkstIlGBZ53lSCWvL8rG6mStQWr1qW1OnnW0hGgzixyszFaihK7P0zKnNMwqnwRdIOr4Ia9PSwLVEy-L1DWckWO0jjMN4ECnES-qOqLPqToKzL8SG9QprA5a1r86bn3ovzZsKjghiW9LdrALbmUUC56PK9jhC42VGsebozDdrIDIU1zIuYI7K25LiUBeK5keKpt6YUJR3uY77-v4xTieLInlDSzqS2x2tAB0ALAiiaJQhxcvF6nJDlwEVhGPXRcKy16JmFoyDBc+RSIPiyC4idQK-hSzC2IlMCmhYf7GAbXcQlCpEgH3A+EDP3bzzAi9GMv0tCUaJ-UPEHuJOAmOVXAmewL7QeyCHujh94yefKX2RxyDghf6t8d84f2jngLa5QQEEDTpRTu3RCB7wPsvGMlkYAlHgc4ZeC4AH5FNPgZeMAVA8DQUvKQK5TSLGXs6bKEwCjzBQhjB+qtcAbhZGZQMnxZikBam1GMr5ZJwBOtdH0JQpCULgNQ2ClhBz0JVmwJhgZy6kDFLJRoKpkASKMnAVRNCFhIJ9Cwo4bDeGkAEcgkoKE1QcNoDowR8B4LACCFEPwNpqCiIAOpHBBL2XARgOjwG0k4Hx8A9xwDsUY2CrVSA8MsSYoRzAWTAHLkYER2UAC0cAZyzCseotuLJVHZMiY0WhoQ-B0zgKItJGS5wlDKLkiAajHGRM0UwngzAUl2IcEEYpSBmoQDCcDKAz5nAxJQT9V0HwPTRJscIlxqT0mzjgA8H4cAtAYFZLU+plgiR8P6YM4w71kHdVTECGAz4el9IdgMwgQypmjPGWE4Z0zgzx12UYB5zAiGH2SeqPhbyUIFDeewgxdCfaqxfKmBcG4lKbhmWI7SRSwiIDMPo7sAByC6jRdHwDAMVKAXzxFKyzowqqMECRwGqcAZFMA0VhMxWEglDC2AvjFFAbo0Knn5BeQC00Op9ZfIBKQHZVy9mtTqfVcx7LLnXNMT9AADPpH6fZCChIAIzzhBqq+FJSWRBngjGAVqspCCqlT6BEh1rimi7Mwekz44A6gXs4Zq+9nD6hYnAS11qXx2qdcYXlML+VuuAGKBwP1MWAtAvSmR7AmnTSRQRcSfoXzKudOsP5cAADM+yfSfD7CAcufLngYGeKadG41JraV1RJINUBFQ7UfIdfaPBGUsngtJK88AzCAgUCGngQbswhNIBCyy9UpCpircwAFUpuW+p5AOVMzBJ0hO4GUWQiBl4TUUgu+1xDqA93zJu71nzqC6CMGYfdB8eWCTgFFO+0js7q3Ek-F+YdmSuUvjtRFgEvrFtvaCx9ocGT-x-lASugIoAQzBKvDELIm4JNbs4aBUhYGmiEFu84idMWoPMAe6omCgOmkAxXKuYGa6QaRSQghyHUPVFIRAsBLdJHGETqI8RtCI3Yzkfo0CoFLFcMiYC4xNy8XaUcVIkFsiqq4AUek5Rmi6n5Nk7pLJmLOOGMmdK8VljaG0rsQ4zZziynZXceJOcATfGptM0EiAtitmcIie9e5Nz4mJK+RU+ZtCROMYU-JldWSumlPKXMzJ5L1neejS0tpLROkIrOZYo1xhfljLMBMn5gmYWucyYsu4Ky1leYaec55QqTC0tYEck5hAYvbIuZym59k7lqdiYaqrhW3mUew0x7K-rfnaQBbw8NP7GUBsHVCoT2jXwItjUBKl6LM1YpxSN4FhLxMfVJcF3hqL+Mzbpf1qNzLWWNYK8akZZgp3te+fqg7wrbhCj8WqOLryatypXQq2QSrSCqqXHADVCw-PasaLqyyQ0YD7Y5c1mxprGrHdtShtrWGD5CBdTw99FhTv+tHcGmbYanrbZfFcmNLDiDxs-Im5NqaM2hsMrm-Nhbv2lr4PNEJshK3Bs+LW7FCUG1NpCa2tgsFO0tOjD2l7nxQlDeHdQNH1auW2v1InJIuA52mjcDwSyid13AFa3DxOu76Qa+cNR2QJ7uVCHh4nNjoKc4a1NJ3ag3coSbyfIQIeI8x4TygFPHec9Yd68Tjb0j697fPg97rmI58sROgoEj3gCSeAQCMKGPkFgIxAA), errors go hasta la vista

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

## Risk

Zero to none, runtime effectively didn't change

## Testing

- Sanity check of the DeFi section

### Engineering

- ☝🏽 

### Operations

- ☝🏽

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3583** <a href="https://app.graphite.dev/github/pr/shapeshift/web/3583" target="_blank">Graphite</a> 
    * **PR #3587** <a href="https://app.graphite.dev/github/pr/shapeshift/web/3587" target="_blank">Graphite</a> 
      * **PR #3598** <a href="https://app.graphite.dev/github/pr/shapeshift/web/3598" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/3598?utm_source=stack-comment).

File: src/state/slices/opportunitiesSlice/resolvers/thorchainsavers/index.ts [hunk 1]
```diff
@@ -133,14 +137,26 @@ export const thorchainSaversStakingOpportunitiesMetadataResolver = async ({
 
     if (!asset || !underlyingAsset || !marketData) continue
 
+    const apy = bnOrZero(
```
Comment (gomesalexandre) [inline src/state/slices/opportunitiesSlice/resolvers/thorchainsavers/index.ts:140]: Reordered to keep tvl and supply next to each other so the comment makes sense

File: src/state/slices/opportunitiesSlice/resolvers/thorchainsavers/index.ts [hunk 2]
```diff
@@ -133,14 +137,26 @@ export const thorchainSaversStakingOpportunitiesMetadataResolver = async ({
...
+    // - supply is the fiat total of assets locked, including the accrued value, accounting in the max. cap
+    const saversSupplyIncludeAccruedFiat = bnOrZero(thorchainPool.savers_depth)
+      .div(bn(10).pow(THOR_PRECISION))
+      .times(marketData.price)
+      .toFixed()
+    const saversMaxSupplyFiat = bnOrZero(thorchainPool.synth_supply)
+      .plus(thorchainPool.synth_supply_remaining)
+      .div(bn(10).pow(THOR_PRECISION))
+      .times(marketData.price)
+      .toFixed()
```
Comment (gomesalexandre) [inline src/state/slices/opportunitiesSlice/resolvers/thorchainsavers/index.ts:158]: Not consumed yet - didn't test whether or not the numbers are correct but needed at type-level now, will revisit when wiring up

File: src/state/slices/opportunitiesSlice/types.ts [hunk 3]
```diff
@@ -12,12 +12,64 @@ export type AssetIdsTuple =
   | readonly [AssetId]
   | readonly []
 
-export type OpportunityMetadata = {
+/**
```
Comment (gomesalexandre) [inline src/state/slices/opportunitiesSlice/types.ts:15]: Copied from the `types` implementation, do we want to export these instead?

File: src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts [hunk 4]
```diff
@@ -126,7 +130,9 @@ export const foxFarmingLpMetadataResolver = async ({
         ] as const,
         name: LP_EARN_OPPORTUNITIES[opportunityId].opportunityName,
       },
-    } as OpportunitiesState[DefiType.LiquidityPool]['byId'],
+    } as Partial<
```
Comment (gomesalexandre) [inline src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts:133]: Taken from `OpportunitiesState[DefiType.LiquidityPool]['byId']`, we might want not to partial it in a follow-up PR, given this effectively loses the opportunitySpecific discrimination: tsc is happy if it's there, but doesn't complain if it isn't. TSC playground confirmed the types are correct, which mean `Partial<>` effectively voids our guarantees here

**Comment by gomesalexandre** [comment]
Opening for early review with a few comments / questions

**Comment by gomesalexandre** [comment]
ser

File: src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts (hunk 4 above)
Comment (0xApotheosis) [reply src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts:132]: I don't think we want to typecast here.

This block should probably be:

```ts
const data: GetOpportunityMetadataOutput<DefiProvider.FoxFarming, DefiType.LiquidityPool> = {
    byId: {
      [opportunityId]: {
        apy,
        assetId: opportunityId,
        provider: DefiProvider.FoxFarming as const,
        tvl,
        type: DefiType.LiquidityPool as const,
        underlyingAssetId: foxEthLpAssetId,
        underlyingAssetIds: foxEthPair,
        underlyingAssetRatios: [
          toBaseUnit(ethPoolRatio.toString(), assets.byId[foxEthPair[0]]?.precision ?? 0),
          toBaseUnit(foxPoolRatio.toString(), assets.byId[foxEthPair[1]]?.precision ?? 0),
        ] as const,
        name: LP_EARN_OPPORTUNITIES[opportunityId].opportunityName,
      },
    },
    type: opportunityType,
  }
```

File: src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts [hunk 5]
```diff
@@ -115,9 +118,9 @@ export const foxFarmingLpMetadataResolver = async ({
       [opportunityId]: {
         apy,
         assetId: opportunityId,
-        provider: DefiProvider.FoxFarming,
+        provider: DefiProvider.FoxFarming as const,
```
Comment (0xApotheosis) [inline src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts:121]: I'm not sure we need these? Typescript is happy without it.

File: src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts [hunk 6]
```diff
@@ -115,9 +118,9 @@ export const foxFarmingLpMetadataResolver = async ({
...
         tvl,
-        type: DefiType.LiquidityPool,
+        type: DefiType.LiquidityPool as const,
```
Comment (0xApotheosis) [inline src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts:123]: I'm not sure we need these? Typescript is happy without it.

File: src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts [hunk 7]
```diff
@@ -218,7 +225,7 @@ export const foxFarmingStakingMetadataResolver = async ({
         name: 'Fox Farming',
         version,
       },
-    } as OpportunitiesState[DefiType.LiquidityPool]['byId'],
+    } as Partial<Record<LpId, OpportunityMetadata<DefiProvider.FoxFarming, DefiType.Staking>>>,
```
Comment (0xApotheosis) [inline src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts:228]: Same here:

```ts
const data: GetOpportunityMetadataOutput<DefiProvider.FoxFarming, DefiType.Staking>
```

And then remove the type casting and `as const`s.

File: src/state/slices/opportunitiesSlice/types.ts (hunk 3 above)
Comment (0xApotheosis) [reply src/state/slices/opportunitiesSlice/types.ts:15]: Yep, I think so. Either that, or move them higher up in the `web` tree - they don't belong in the opportunities slice.

**Review by 0xApotheosis — COMMENTED** [review]
Some wild TypeScript fun, here. Added some thoughts. 

As far as I can tell the types work without casting the `DefiProvider` and `DefiType` enums as consts - what have I missed?

File: src/state/slices/opportunitiesSlice/types.ts (hunk 3 above)
Comment (gomesalexandre) [reply src/state/slices/opportunitiesSlice/types.ts]: 3f3b801c5e0ae53d18fb8a532c443d8a4c905fd9

File: src/state/slices/opportunitiesSlice/resolvers/idle/constants.ts [hunk 8]
```diff
@@ -48,7 +48,7 @@ export const BASE_OPPORTUNITIES_BY_ID: PartialRecord<
   },
   'eip155:1/erc20:0xf34842d05a1c888ca02769a633df37177415c2f8': {
     assetId: 'eip155:1/erc20:0xf34842d05a1c888ca02769a633df37177415c2f8',
-    provider: DefiProvider.Idle,
+    provider: DefiProvider.Idle as const,
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/state/slices/opportunitiesSlice/resolvers/idle/constants.ts]: Disregard me: 3a39e20ea138319032420d79d970c7900e3c167f

File: src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts (hunk 4 above)
Comment (gomesalexandre) [reply src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts]: Much better, ty! No assertion, no partial and properly leveraging the newly added generics 🤌 
3a39e20ea138319032420d79d970c7900e3c167f

File: src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts (hunk 5 above)
Comment (gomesalexandre) [reply src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts]: 3a39e20ea138319032420d79d970c7900e3c167f

File: src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts (hunk 6 above)
Comment (gomesalexandre) [reply src/state/slices/opportunitiesSlice/resolvers/foxFarming/index.ts]: 3a39e20ea138319032420d79d970c7900e3c167f

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**